internal/policies/client.go

```diff
@@ -0,0 +1,330 @@
+	if err != nil {
+		return nil, err
+	}
+	opts := client.ListOptions{
+		LabelSelector: labelSelector,
+	}
+	err = f.client.List(ctx, namespaceList, &opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return namespaceList.Items, nil
+}
+
+func (f *Client) getAdmissionPolicies(ctx context.Context, namespace string) ([]policiesv1.AdmissionPolicy, error) {
+	policies := &policiesv1.AdmissionPolicyList{}
+	err := f.client.List(ctx, policies, &client.ListOptions{Namespace: namespace})
+	if err != nil {
+		return nil, err
+	}
+
+	return policies.Items, nil
+}
+
+// groupPoliciesByGVRAndLabelSelectorg roups policies by GVR and LabelSelector.
+// If namespaced is true, it will skip cluster-wide resources, otherwise it will skip namespaced resources.
+func (f *Client) groupPoliciesByGVRAndLabelSelector(ctx context.Context, policies []policiesv1.Policy, namespaced bool) (map[schema.GroupVersionResource]map[string][]*Policy, error) {
+	resources := make(map[schema.GroupVersionResource]map[string][]*Policy)
+	for _, policy := range policies {
+		url, err := f.getPolicyServerURLRunningPolicy(ctx, policy)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, rules := range policy.GetRules() {
+			for _, resource := range rules.Resources {
+				for _, version := range rules.APIVersions {
+					for _, group := range rules.APIGroups {
+						gvr := schema.GroupVersionResource{
+							Group:    group,
+							Version:  version,
+							Resource: resource,
+						}
+						isNamespaced, err := f.isNamespacedResource(gvr)
+						if err != nil {
+							return nil, err
+						}
+						if namespaced && !isNamespaced {
+							// continue if resource is clusterwide
+							continue
+						}
+						if !namespaced && isNamespaced {
+							// continue if resource is namespaced
+							continue
+						}
+						selector, err := metav1.LabelSelectorAsSelector(policy.GetObjectSelector())
+						if err != nil {
+							return nil, err
+						}
+						labelSelector := selector.String()
+
+						policy := Policy{
+							Policy:       policy,
+							PolicyServer: url,
+						}
+
+						addPolicyToMap(resources, gvr, labelSelector, &policy)
+					}
+				}
+			}
+		}
+	}
+
+	return resources, nil
+}
+
+func addPolicyToMap(resources map[schema.GroupVersionResource]map[string][]*Policy, gvr schema.GroupVersionResource, labelSelector string, policy *Policy) {
+	if _, ok := resources[gvr]; !ok {
+		resources[gvr] = map[string][]*Policy{
+			labelSelector: {policy},
+		}
+		return
+	}
+
+	if _, ok := resources[gvr][labelSelector]; !ok {
+		resources[gvr][labelSelector] = []*Policy{policy}
+	} else {
+		resources[gvr][labelSelector] = append(resources[gvr][labelSelector], policy)
+	}
+}
+
+// Method to check if the given resource is namespaced or not.
+func (f *Client) isNamespacedResource(gvr schema.GroupVersionResource) (bool, error) {
+	gvk, err := f.client.RESTMapper().KindFor(gvr)
```

Comment:
AFAIK I don't think we are hitting the API server here since the rest mapper is initialized when the client is created. I will double-check this, but you can take a look here: https://github.com/kubernetes-sigs/controller-runtime/blob/56159419231e985c091ef3e7a8a3dee40ddf1d73/pkg/client/apiutil/restmapper.go#L35